src/iterators/chunked_read_ahead_iterator.rs

```diff
@@ -0,0 +1,368 @@
+                if let Ok(result) = r.recv() {
+                    result
+                } else {
+                    return None;
+                }
+            } else {
+                let box_message: Box<dyn Any + Send> =
+                    Box::new("Reciever is None yet iteration is ongoing.");
+                Err(box_message)
+            };
+            // If the new chunk is present and Ok, convert it to an iterator, store it on ``self``,
+            // and return its next value ( shutting down our reciever if the next value is None).
+            // if the new chunk is an Err, raise it to the main thread.
+            match res_r {
+                Ok(next_chunk) => {
+                    self.current_chunk = next_chunk.into_iter();
+                    if let Some(result) = self.current_chunk.next() {
+                        Some(result)
+                    } else {
+                        self.receiver = None;
+                        None
+                    }
+                }
+                Err(e) => {
+                    self.receiver = None;
+                    resume_unwind(e);
+                }
+            }
+        }
+    }
+}
+
+/// Trait that implements ``read_ahead`` a method for converting ``self`` to a
+/// ``ChunkedReadAheadIterator``.
+#[allow(clippy::module_name_repetitions)]
+pub trait IntoChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    /// Converts a struct into a ``ChunkedReadAheadIterator``, with chunks of size `chunk_size`
+    /// and a read ahead buffer of `num_chunk_buffer_size` chunks.
+    fn read_ahead(
+        self,
+        chunk_size: usize,
+        num_chunk_buffer_size: usize,
+    ) -> ChunkedReadAheadIterator<T>
+    where
+        Self: Send + 'static;
+}
+
+impl<I, T> IntoChunkedReadAheadIterator<T> for I
+where
+    T: Send + 'static,
+    I: Iterator<Item = T>,
+{
+    fn read_ahead(
+        self,
+        chunk_size: usize,
+        num_chunk_buffer_size: usize,
+    ) -> ChunkedReadAheadIterator<T>
+    where
+        Self: Send + 'static,
+    {
+        ChunkedReadAheadIterator::new(self, chunk_size, num_chunk_buffer_size)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rstest::rstest;
+    use std::mem::drop;
+    use std::panic;
+    // use std::panic;
+    use std::thread::sleep;
+    use std::time::Duration;
+
+    #[rstest]
+    #[case(1)] // smallest possible
+    #[case(2)]
+    #[case(4)]
+    #[case(8)]
+    #[case(16)] // larger than the inner iterator
+    fn test_wrapping_empty_iter(#[case] chunk_size: usize) {
+        let test_vec: Vec<usize> = Vec::new();
+
+        let mut chunked_iter = test_vec.into_iter().read_ahead(chunk_size, 1); // TODO add into chunked readahead
+        assert_eq!(chunked_iter.next(), None);
+    }
+
+    #[rstest]
+    #[case(1)] // smallest possible
+    #[case(2)]
+    #[case(4)]
+    #[case(8)]
+    #[case(16)] // larger than the inner iterator
```

Comment:
Comment doesn't apply to this case, right?
Should we also parameterize over the number of chunks?